### PR TITLE
[PluggableDevice] Add TF_AddNVariant and TF_ZerosLikeVariant

### PR DESCRIPTION
**Pull request by PatriceVignola**
Support for TF_VARIANT is not really available for pluggable devices through the API since it's very hard (or even impossible) to read the content of a C++ object through ABIs. Even if we were to use the exact same headers as TensorFlow uses for all C++ objects, we would most likely need to use the exact same compiler.

Since some models depend heavily on variant tensors which contain `TensorList` objects, we'd like to propose 2 new APIs that address this issue:

```cpp
TF_CAPI_EXPORT extern void TF_AddNVariant(
    TF_OpKernelContext* ctx,
    void (*binaryAddFunc)(TF_OpKernelContext* ctx, const TF_Tensor* a, const TF_Tensor* b, TF_Tensor* out),
    TF_Status* status);

TF_CAPI_EXPORT extern void TF_ZerosLikeVariant(
    TF_OpKernelContext* ctx,
    void (*zerosLikeFunc)(TF_OpKernelContext* ctx, const TF_Tensor* input, TF_Tensor* out),
    TF_Status* status);
```

Like the `TF_AssignVariable` and `TF_AssignUpdateVariable` functions currently available in `kernels_experimental.h`, the goal of `TF_AddNVariant` and `TF_ZerosLikeVariant` is to allow plugins to implement those 2 key operations by treating the Variant objects as a black box. The Variant objects are unwrapped within TensorFlow core, which then call the binary_add_func or zeros_like_func callbacks provided by the user, which are guaranteed to contain tensors of primitives (e.g. `TF_FLOAT`).

### CODE REVIEW

File: tensorflow/core/framework/variant_ops.cc [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (PatriceVignola) [reply tensorflow/core/framework/variant_ops.cc]: Not really, I just put it there by default. I moved them to kernels.

**Review by mdanatg — COMMENTED** [review]
I need to have a closer look at the low level code before I can give better feedback since I'm not too familiar with it, but here are a couple of comments meanwhile.

File: tensorflow/c/kernels_experimental.cc [hunk 2]
```diff
@@ -391,3 +396,217 @@ bool TF_IsRefInput(TF_OpKernelContext* ctx, int i, TF_Status* status) {
...
+                                Variant* out) {
+    DCHECK_NE(out, nullptr);
+    if (a.TypeId() != b.TypeId()) {
+      return ::tensorflow::errors::Internal(
+          "BinaryOpVariants: Variants a and b have different "
+          "type ids.  Type names: '",
+          a.TypeName(), "' vs. '", b.TypeName(), "'");
+    }
+
+    if (a.TypeId() == tensorflow::TypeIndex::Make<::tensorflow::TensorList>()) {
```
[the remark this answers is not in the document]
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.cc:480]: Doesn't the registry need to know the registration at compile time, including the name of the device? I'm wondering how we could leverage this in the context of pluggable devices.

File: tensorflow/c/kernels_experimental.cc [hunk 3]
```diff
@@ -391,3 +396,217 @@ bool TF_IsRefInput(TF_OpKernelContext* ctx, int i, TF_Status* status) {
   TF_SetStatus(status, TF_OK, "");
   return cc_ctx->input_is_ref(i);
 }
+
+template <typename T>
+static Status ValidateVariantType(const Variant& variant) {
+  if (variant.get<::tensorflow::TensorList>() == nullptr) {
```
[the remark this answers is not in the document]
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.cc]: Good catch, thank you!

File: tensorflow/c/kernels_experimental.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdanatg) [reply tensorflow/c/kernels_experimental.cc:480]: The registry is just a static variable that can be modified like any other. Though I suspect we'd need to make it thread safe to allow dynamic modifications.
Even if done during initialization like the existing registrations, I think the device name would be DEVICE_DEFAULT, like the kernel registry does.
@penpornk would this make sense or are there other details we should be aware of?

[the remark this answers is not in the document]
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.cc:480]: I don't know that area of the code very well so bear with me, but it looks like everything that interacts with the registry is providing the actual implementation during initialization. In our case, even if using DEVICE_DEFAULT works, plugins need to provide their own implementation of the operations to run on the unwrapped tensors (e.g. Add for AddNVariant). I'm also not sure how well DEVICE_DEFAULT would work when dealing with multiple pluggable devices that all have their own unique device.

Also, how does the registry handle overwriting existing registrations? For example, plugins are allowed to name their device "GPU", but there's already a "GPU" registration for ZerosLikeVariant in the registry.

File: tensorflow/c/kernels_experimental.cc [hunk 4]
```diff
@@ -391,3 +396,217 @@ bool TF_IsRefInput(TF_OpKernelContext* ctx, int i, TF_Status* status) {
...
+        }
+
+        auto* ctx = reinterpret_cast<TF_OpKernelContext*>(cc_ctx);
+        zeros_like_func(ctx, input, out);
+      }
+    }
+    return cc_ctx->status();
+  };
+
+  DCHECK_NE(out, nullptr);
```
Comment (mihaimaruseac) [inline tensorflow/c/kernels_experimental.cc:559]: Can you replace all checks (debug or not) with status returns, please?

We have an internal presubmit checker that no longer allows checks due to the fact that is causes downtime / offers a chance for denial of service.

Debug checks are slightly worse as they are only active during debug builds but we rarely have them.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by mihaimaruseac** [comment]
> cc @mihaimaruseac for a security assessment into the ability to inject arbitrary code into the computation.

Seems ok as the injection possibility is limited.

File: tensorflow/c/kernels_experimental.cc (hunk 4 above)
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.cc]: @mihaimaruseac Sorry for the delay, I replaced the `CHECKs` with status returns.

File: tensorflow/c/kernels_experimental.cc [hunk 5]
```diff
@@ -391,3 +396,217 @@ bool TF_IsRefInput(TF_OpKernelContext* ctx, int i, TF_Status* status) {
...
+    return ::tensorflow::errors::Internal(
+        "VariantBinaryOpFn: Could not access object 'a', type_index: ",
+        type_index_name);
+  }
+
+  return Status::OK();
+}
+
+void TF_AddNVariant(TF_OpKernelContext* ctx,
+                    void (*binary_add_func)(TF_OpKernelContext* ctx,
```
[the remark this answers is not in the document]
Comment (cantonios) [reply tensorflow/c/kernels_experimental.cc:415]: Can you expand on this? AFAIK the `std::function` ABI *should* be stable for post c++11-fully-compatible compilers (apart from perhaps compiler bugs).  Or are you talking about cross between `_GLIBCXX_USE_CXX11_ABI=0` and `_GLIBCXX_USE_CXX11_ABI=1`?

[the remark this answers is not in the document]
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.cc:415]: Both. My experience was that compilers have a lot of issues (i.e. bugs) with cross-abi compatibility when it comes to even simple C++ types like std::function. But it's also the fact that this is a C API, so I don't think C++ types should be passed through it.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by PatriceVignola** [comment]
Is there anything I can do to satisfy the failing checks or are they infra issues? They don't seem related to my changes.

**Comment by gbaned** [comment]
> Is there anything I can do to satisfy the failing checks or are they infra issues? They don't seem related to my changes.

Hi @PatriceVignola  Nothing pending from your end, this PR is processing internally. Thank you!

**Comment by PatriceVignola** [comment]
@gbaned Is there any update on this PR? I still see some failed checks. Are they supposed to re-run automatically? Thanks!

**Comment by PatriceVignola** [comment]
@gbaned Ah, it looks like a commit was made after I resolved the conflicts yesterday. Should be all good now!

**Comment by PatriceVignola** [comment]
@gbaned Ah, it looks like a commit was made after I resolved the conflicts yesterday. Should be all good now!

**Review by aaudiber — CHANGES_REQUESTED** [review]
Please fix the license formatting and copyright date for the newly added files

**Review by aaudiber — APPROVED** [review]

**Review by aaudiber — APPROVED** [review]

**Review by aaudiber — APPROVED** [review]

**Comment by PatriceVignola** [comment]
@gbaned is there anything to do on my end to fix the feedback/copybara checks? I can't see the details.

File: tensorflow/core/kernels/data/optional_ops_util.cc [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+namespace tensorflow {
+namespace data {
+
+Status OptionalZerosLike(OpKernelContext* ctx, const OptionalVariant& x,
+                         OptionalVariant* y,
+                         std::function<Status(OpKernelContext* ctx,
+                                              const Tensor& input, Tensor* out)>
+                             zeros_like_func) {
+  if (!x.has_value()) {
+    *y = x;
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/data/optional_ops_util.cc:31]: [Internal mutation testing](https://testing.googleblog.com/2021/04/mutation-testing.html) detected that all existing tests would pass if this line is removed. Can you write a test that would check that `*y` gets the value of `x` in this scenario, so that accidentally removing this line would be caught by CI, please?

File: tensorflow/core/kernels/data/optional_ops_util.cc [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    std::function<Status(OpKernelContext* ctx, const Tensor& a, const Tensor& b,
+                         Tensor* out)>
+        binary_add_func) {
+  // TODO(skyewm): should adding a value to a non-value be a no-op instead?
+  if (a.has_value() != b.has_value()) {
+    return errors::InvalidArgument(
+        "Cannot add optionals because one has a value and the other doesn't.");
+  }
+  if (!a.has_value()) {
+    *out = a;
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/data/optional_ops_util.cc:56]: [Internal mutation testing](https://testing.googleblog.com/2021/04/mutation-testing.html) detected that all existing tests would pass if this line is removed. Can you write a test that would check that `*y` gets the value of `x` in this scenario, so that accidentally removing this line would be caught by CI, please?

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by mihaimaruseac** [comment]
Also, ClangTidy suggestions (line numbers might not be exact due to internal import)
	
```
  .../tensorflow/core/kernels/data/optional_ops_util.h:39 ClangTidy: missing #include <utility> for 'std::move'
  .../tensorflow/core/kernels/variant_ops_util.cc:39 ClangTidy: missing #include <utility> for 'std::move'
  .../tensorflow/core/kernels/variant_ops_util.cc:35 ClangTidy: missing #include <functional> for 'std::function'
  .../tensorflow/core/kernels/data/optional_ops_util.h:100 ClangTidy: missing #include <functional> for 'std::function'
  .../tensorflow/core/kernels/tensor_list_util.cc:29 ClangTidy: missing #include <functional> for 'std::function'
```

**Comment by mihaimaruseac** [comment]
Finally,

```
comparison of different enumeration types ('tensorflow::DataType' and 'TF_DataType') [-Werror,-Wenum-compare]
    if (cc_a.dtype() == TF_VARIANT) {
        ~~~~~~~~~~~~ ^  ~~~~~~~~~~
1 error generated.
```

File: tensorflow/core/kernels/data/optional_ops_util.cc (hunk 6 above)
Comment (PatriceVignola) [reply tensorflow/core/kernels/data/optional_ops_util.cc:31]: It's true, this line isn't necessary because the caller functions already initialize the output to an empty optional beforehand. I can just remove the line.

File: tensorflow/core/kernels/data/optional_ops_util.cc (hunk 7 above)
Comment (PatriceVignola) [reply tensorflow/core/kernels/data/optional_ops_util.cc:56]: It's true, this line isn't necessary because the caller functions already initialize the output to an empty optional beforehand. I can just remove the line.

**Comment by PatriceVignola** [comment]
@mihaimaruseac Anything else I should change to satisfy the internal checks?

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
It needs to be imported again, sorry, I didn't get a chance to look at it before (since I'm no longer in TF, reviewing TF PRs is a lower priority for me)

**Comment by mihaimaruseac** [comment]
Still seeing ClangTidy complaints:

```
.../tensorflow/c/kernels_experimental.cc:446: missing #include <string> for 'std::string'
.../tensorflow/core/kernels/data/optional_ops_util.cc:27: missing #include <functional> for 'std::function'
.../tensorflow/core/kernels/data/optional_ops_util.cc:33: missing #include <vector> for 'std::vector'
.../tensorflow/core/kernels/data/optional_ops_util.cc:37: missing #include <utility> for 'std::move'
.../tensorflow/core/kernels/data/optional_ops_util.h:41: missing #include <memory> for 'std::make_shared'
.../tensorflow/core/kernels/variant_ops_util.h:27: function 'tensorflow::AddNVariant' has a definition with different parameter names
```

Also, build currently seems broken due to

```
.../tensorflow/c/kernels_experimental.cc:32:10: error: module //.../tensorflow/c:kernels_experimental does not depend on a module exporting '.../tensorflow/core/platform/abi.h'; to fix run:
	build_cleaner //.../tensorflow/c:kernels_experimental
#include ".../tensorflow/core/platform/abi.h"
```

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
Please make sure all header inclusions are also paired with the corresponding BUILD rules. Internal tests are now failing because of

```
.../tensorflow/c/kernels_experimental.cc:30:10: error: module //.../tensorflow/c:kernels_experimental does not depend on a module exporting '.../tensorflow/core/kernels/tensor_list.h'; to fix run:
	build_cleaner //.../tensorflow/c:kernels_experimental
#include ".../tensorflow/core/kernels/tensor_list.h"
```

**Comment by PatriceVignola** [comment]
@mihaimaruseac Ah yes, the commit I pushed yesterday was just the first part of the fixes. My local builds for the include fixes didn't have time to complete so I didn't push the rest.

Also, is there a script or command line I should use to run clang-tidy the same way it's run internally, so that I can catch those errors earlier? I couldn't make it to work with all the necessary includes and it seemed relatively error-prone.

Finally, can I know the command that the internal tests are running to reproduce the include errors you mentioned? I believed I fixed them, but I want to cover all bases before pushing yet another commit that may break internal tests.

Thank you!

**Comment by angerson** [comment]
>Finally, can I know the command that the internal tests are running to reproduce the include errors you mentioned? I believed I fixed them, but I want to cover all bases before pushing yet another commit that may break internal tests.

This type of failure is new to me, but it seems to have been found via the internal-version-of-bazel's-crosstool's `--features=layering_check` flag on `bazel build` -- maybe you could try that and see what happens? I'm not sure if it will work and can't really find any docs about it.

>Also, is there a script or command line I should use to run clang-tidy the same way it's run internally, so that I can catch those errors earlier? I couldn't make it to work with all the necessary includes and it seemed relatively error-prone.

I've looked into this before: it's tricky because changes to TF are blocked by ClangTidy rules that are common to the whole of Google, rather than something we configured, and there are a lot of custom rules that aren't available externally. So if we introduce a clangtidy presubmit, it won't exactly match the internal rules and we'll probably have to use a subset to make it work. @learning-to-play this would be valuable but would take some time to find out if it would work.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
Still errors

```
(18:20:29) ERROR: .../tensorflow/core/kernels/BUILD:7011:11: Compiling .../tensorflow/core/kernels/tensor_list.cc failed: missing input file '//.../tensorflow/core/kernels:tensor_list.cc'
(18:20:29) ERROR: .../tensorflow/core/kernels/BUILD:7011:11: Compiling .../tensorflow/core/kernels/tensor_list.cc failed: 1 input file(s) do not exist
(18:20:29) ERROR: .../tensorflow/core/kernels/BUILD:7011:11: Compiling .../tensorflow/core/kernels/tensor_list.cc failed: missing input file '//.../tensorflow/core/kernels:tensor_list.cc'
(18:20:29) ERROR: .../tensorflow/core/kernels/BUILD:7011:11: Compiling .../tensorflow/core/kernels/tensor_list.cc failed: 1 input file(s) do not exist
(18:20:33) ERROR: <internal_path>: Linking <internal_path>.so failed: (Exit 1) clang failed: error executing command[...]
ld.lld: error: undefined symbol: tensorflow::data::OptionalZerosLike(tensorflow::OpKernelContext*, tensorflow::data::OptionalVariant const&, tensorflow::data::OptionalVariant*, std::__ndk1::function<tensorflow::Status (tensorflow::OpKernelContext*, tensorflow::Tensor const&, tensorflow::Tensor*)>)
....
```

You are still having a `tensor_list.cc` forgotten in `core/kernels/BUILD`. Unsure if `tensor_list.cc` can actually be moved to a different directory, but let's keep trying this way.

This is not detected in OSS runs because the tests are marked as `manual`, so they only run (and are compiled) if specifically requested on the command line.

**Comment by PatriceVignola** [comment]
@mihaimaruseac I moved it to another directory because an Android OSS test was failing since it doesn't depend on kernels.

I'll remove `tensor_list.cc`, but obviously I won't know if it breaks something else by just running the OSS tests or fixes anything. What command can I run that could reproduce the errors above? I apologize for all the back and forth, I'm running the test suites that I'm aware of locally but it seems like there are a lot more tests than what the CI shows publicly.

**Review by penpornk — COMMENTED** [review]
Thank you for the change! I think the previous error is solved now. 
There are still other errors though:
```
ld.lld: error: undefined symbol: tensorflow::AddNVariant(tensorflow::OpKernelContext*, std::__ndk1::function<tensorflow::Status (tensorflow::OpKernelContext*, tensorflow::Variant const&, tensorflow::Variant const&, tensorflow::Variant*)>)
>>> referenced by aggregate_ops.cc
>>>               blaze-out/.../tensorflow/core/kernels/_objs/portable_tensorflow_kernels/aggregate_ops.pic.o:(tensorflow::AddNOp<Eigen::ThreadPoolDevice, tensorflow::Variant>::Compute(tensorflow::OpKernelContext*))

ld.lld: error: undefined symbol: tensorflow::TensorList::~TensorList()
>>> referenced by list_kernels.cc
>>>               blaze-out/.../tensorflow/core/kernels/_objs/portable_tensorflow_kernels/list_kernels.pic.o:(tensorflow::ForwardInputOrCreateNewList(tensorflow::OpKernelContext*, int, int, tensorflow::TensorList const&, tensorflow::TensorList**))
>>> referenced by list_kernels.cc
>>>               blaze-out/.../tensorflow/core/kernels/_objs/portable_tensorflow_kernels/list_kernels.pic.o:(tensorflow::EmptyTensorList::Compute(tensorflow::OpKernelContext*))
>>> referenced by list_kernels.cc
>>>               blaze-out/.../tensorflow/core/kernels/_objs/portable_tensorflow_kernels/list_kernels.pic.o:(tensorflow::TensorListReserve::Compute(tensorflow::OpKernelContext*))
>>> referenced 96 more times
...
```

Here is the command to reproduce:
```
bazel test --config=opt //tensorflow/lite/tools/benchmark/android:libtensorflowlite_benchmark_plus_flex.so --test_env=TF2_BEHAVIOR=1
```
(I haven't tried it yet. You might also need to sync to a commit before 044fc36bf4490a9b88b8ceaa0d6f227c71aad0fb, which broke some builds, e.g., sync to fc9ac5879be3cd93bb04f95341dbf4d165743b01)

**Comment by PatriceVignola** [comment]
@penpornk Thanks. I had to sync to an older commit since LLVM kept breaking. But I still cannot reproduce it locally:

```sh
bazel test --config=opt //tensorflow/lite/tools/benchmark/android:libtensorflowlite_benchmark_plus_flex.so --test_env=TF2_BEHAVIOR=1
```

```bazel
...
INFO: Found 1 target and 0 test targets...
Target //tensorflow/lite/tools/benchmark/android:libtensorflowlite_benchmark_plus_flex.so up-to-date:
  bazel-bin/tensorflow/lite/tools/benchmark/android/libtensorflowlite_benchmark_plus_flex.so
INFO: Elapsed time: 0.281s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
```

Nevertheless, I made a commit where I added the missing `.cc` files to the `mobile_srcs_no_runtime` filegroup, but it's weird that I cannot reproduce the failure by running the command that you gave me, even without this change.

**Comment by penpornk** [comment]
You're right. I can't reproduce the error in our docker environment either. (The failure was from our internal test environment.) :'(

I tried your new fix and get this error:
```bazel
$ bazel test --config=opt //third_party/tensorflow/lite/tools/benchmark/android:benchmark_model_plus_flex --test_env=TF2_BEHAVIOR=1
ERROR: //tensorflow/core/BUILD:1104:11: Compiling //tensorflow/core/framework/tensor_list_util.cc failed: (Exit 1) clang failed: error executing command (from target //tensorflow/core:portable_tensorflow_lib_lite_no_runtime) 
...
//tensorflow/core/framework/tensor_list_util.cc:20:10: fatal error: '//tensorflow/core/framework/op_kernel.h' file not found
#include "tensorflow/core/framework/op_kernel.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target //tensorflow/lite/tools/benchmark/android:benchmark_model_plus_flex failed to build
FAILED: Build did NOT complete successfully
```

I don't think we should keep adding files to the list. There are too many dependencies, e.g., adding `op_kernel.h` gives
```bazel
In file included from third_party/tensorflow/core/framework/tensor_list_util.cc:20:
./third_party/tensorflow/core/framework/op_kernel.h:28:10: fatal error: 'third_party/tensorflow/core/framework/cancellation.h' file not found
#include "third_party/tensorflow/core/framework/cancellation.h"
1 error generated.
```

Adding `cancellation.h` gives
```bazel
./third_party/tensorflow/core/framework/cancellation.h:22:10: fatal error: 'third_party/tensorflow/core/lib/core/notification.h' file not found
#include "third_party/tensorflow/core/lib/core/notification.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
And so on.

Sorry if this has been brought up before: Why do we move tensor_list.cc from kernels/ to framework/ ?

Is it possible to keep tensor_list.cc and tensor_list.h untouched, and guard all new files away from mobile platforms?

**Comment by PatriceVignola** [comment]
I originally moved the files because the "Android Demo" check couldn't find them. This target was somehow depending on `kernels_experimental`, which in turn needs those files. Are there any reasons that Android build should depend on `kernels_experimental` at all, assuming this library's purpose is to be used by pluggable devices?

I'm going to try something else locally to fix this "Android Demo" build error, but I'm flying relatively blind if I can't repro the other failures locally :/

**Comment by penpornk** [comment]
> I originally moved the files because the "Android Demo" check couldn't find them. This target was somehow depending on kernels_experimental, which in turn needs those files. Are there any reasons that Android build should depend on kernels_experimental at all, assuming this library's purpose is to be used by pluggable devices?

No, I don't think Android build should depend on `kernels_experimental`. It could be an automatic dependency through `kernels_hdrs`:
https://github.com/tensorflow/tensorflow/blob/9e938054241976368f4614f96070722f9181984c/tensorflow/c/BUILD#L591-L596

We could do a `select` similar to here:
https://github.com/tensorflow/tensorflow/blob/9e938054241976368f4614f96070722f9181984c/tensorflow/c/BUILD#L621-L636

E.g.,
```bazel
    name = "kernels_hdrs",
    hdrs = ["kernels.h"] + select({
        "//tensorflow:mobile": [],
        "//conditions:default": ["kernels_experimental.h"],
    }),
```
Or be even stricter since PluggableDevice is only meant for / tested on Linux x86, Windows, and Mac.
```bazel
load("@bazel_skylib//lib:selects.bzl", "selects")
...
    name = "kernels_hdrs",
    hdrs = ["kernels.h"] + selects.with_or({
        # I'm just listing them in alphabetical order :)
        ("//tensorflow:linux_x86_64", "//tensorflow:macos", "//tensorflow:windows"): ["kernels_experimental.h"],
        "//conditions:default": [],
    }),
```

Can we move everything back to `tensorflow/core/kernels` and try using selects as necessary to exclude whatever `srcs`, `hdrs`, `deps` from Android (so that Android Demo don't need to find them in the first place)?

**Comment by PatriceVignola** [comment]
@penpornk I found the problem: the android demo somehow doesn't even use the BUILD file to compile. I renamed `kernels_experimental.cc` to `dummy.cc` without changing the name in any of the BUILD files, and the android demo target was somehow trying to compile it. There's probably a script or a rule somewhere that tries to compile all files within a directory, but I don't know where it is.

Therefore, I can workaround it by placing all the new code within a `#ifndef IS_MOBILE_PLATFORM`. Does that seem like a good solution? I really wasn't able to find the script that tries to compile all files within that directory, but it feels weird that such a script would even exist.

**Comment by penpornk** [comment]
@PatriceVignola Thank you for investigating! Yes, I think that's a good workaround. :) (And I can look more into Android Demo later after the branch cut.)

**Comment by PatriceVignola** [comment]
@penpornk I pushed the changes. Hopefully everything will pass this time :)

File: tensorflow/core/kernels/tensor_list.cc [hunk 8]
```diff
@@ -51,32 +51,6 @@ void TensorList::Encode(VariantTensorData* data) const {
...
-#define REGISTER_LIST_COPY(DIRECTION)                                         \
-  INTERNAL_REGISTER_UNARY_VARIANT_DEVICE_COPY_FUNCTION(TensorList, DIRECTION, \
-                                                       TensorListDeviceCopy)
-
-REGISTER_LIST_COPY(VariantDeviceCopyDirection::HOST_TO_DEVICE);
-REGISTER_LIST_COPY(VariantDeviceCopyDirection::DEVICE_TO_HOST);
-REGISTER_LIST_COPY(VariantDeviceCopyDirection::DEVICE_TO_DEVICE);
-
-REGISTER_UNARY_VARIANT_DECODE_FUNCTION(TensorList, TensorList::kTypeName);
-
```
Comment (penpornk) [inline tensorflow/core/kernels/tensor_list.cc:79]: Just curious, why is this moved to list_kernels.cc?

**Review by penpornk — APPROVED** [review]
That was fast! Thank you! Let's see how this goes. :)

File: tensorflow/core/kernels/tensor_list.cc (hunk 8 above)
Comment (PatriceVignola) [reply tensorflow/core/kernels/tensor_list.cc:79]: There was a duplicate registration at build time when `kernels_experimental` was trying to depend on `tensor_list`. `kernels_experimental` needs the `TensorList`  class and functions, but it doesn't need to depend on the registration stuff.

**Review by penpornk — CHANGES_REQUESTED** [review]
I have some more build fixes (and a few extra nits). The previously failed tests are good and I'm running more tests now. :)

**Comment by PatriceVignola** [comment]
@penpornk Thanks for this! I pushed the changes.

**Review by penpornk — APPROVED** [review]
Thank you for the fixes!
